File: 230- Kth Smallest Element in a BST-二叉搜索树中第K小的元素/230- Kth Smallest Element in a BST-二叉搜索树中第K小的元素.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
// var kthSmallest = function(root, k) {
//     let list = []
//     function midTravel(tree){
//         if(!tree || list.length === k){
//             return 
//         }
//         midTravel(tree.left)
//         list.push(tree.val)
//         midTravel(tree.right)
//     }
//     midTravel(root)
//     return list[k - 1]
// };

var kthSmallest = function(root, k) {
    let t = k
    let res = 0
    function midTravel(tree){
        if(!tree || t === 0){
            return 
        }
        midTravel(tree.left)
        t--
        if(t === 0){
            res = tree.val
        }
        midTravel(tree.right)
    }
    midTravel(root)
    return res
};

File: 230- Kth Smallest Element in a BST-二叉搜索树中第K小的元素/230- Kth Smallest Element in a BST-二叉搜索树中第K小的元素.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// class Solution {
//     ArrayList<Integer> list = new ArrayList<Integer>();
//     public int kthSmallest(TreeNode root, int k) {
//         midOrder(root);
//         Collections.sort(list);
//         return list.get(k - 1);
//     }

//     public void midOrder(TreeNode tree) {
//         if(tree == null) {
//             return;
//         }
//         list.add(tree.val);
//         midOrder(tree.left);
//         midOrder(tree.right);
// 	}
// }

class Solution {
    TreeNode tem;
    int t,res;
    public int kthSmallest(TreeNode root, int k) {
        t = k;
        tem = root;
        dfs(tem);
        return res;
    }
    public void dfs(TreeNode tem){
        if(tem == null) return;
        dfs(tem.left);
        t--;
        if(t == 0) res = tem.val;
        dfs(tem.right);
    }
}
